doc/plutip-testing.md

```diff
@@ -26,13 +27,43 @@ CTL depends on a number of binaries in the `$PATH` to execute Plutip tests:
+CTL can help you test the offchain `Contract`s from your project (and consequently the interaction of onchain and offchain code) by spinning up a disposable private testnet via Plutip and making all your `Contract`s interact with it.
+
+There are two approaches to writing such tests.
+
+[First](#testing-in-aff-context) is to use the `Contract.Test.Plutip.runPlutipContract` function, which takes a single `Contract`, launches a Plutip cluster and executes the passed contract.
+This function runs in `Aff`; it will also throw an exception should contract fail for any reason.
+After the contract execution the Plutip cluster is terminated.
+You can either call it directly from your test's main or use any library for grouping and describing tests which support effects in the test body, like Mote.
+
+[Mote](https://github.com/garyb/purescript-mote) is a DSL for defining and grouping tests (plus other quality of life features, e.g. skipping marked tests).
+
+
+[Second](#testing-with-mote) (and more widely used) approach is to first build a tree of tests (in CTL's case a tree of `ContractTest` types -- basically a function from some distribution of funds to a `Contract a`) via Mote and then use the `Contract.Test.Plutip.testPlutipContracts` function to execute them.
+This allows to set up a Plutip cluster only once per top-level groups and tests passed to the `testPlutipContracts` and then use it in many independent tests.
+The function will interpret a `MoteT` (effectful test tree) into `Aff`, which you can then actually run.
+
+The [`ctl-scaffold` template](../templates/ctl-scaffold) provides a simple `Mote`-based example.
+
+
+CTL will run contracts in your test bodies and will print errors for any failed tests.
+Only test body failures are checked and this works fine if you want to make sure your `Contract`s execute without errors; if you want to add more precise checks (like checking that particular token is now at some address, that some exact amount was transferred, etc.) then you need to either write these checks manually in a `Contract` monad and then throw errors, or (preferably) use the [assertions library](./test-utils.md).
+
+The communication with Plutip happens via the `plutip-server`'s HTTP interface, which allows to start or stop a cluster.
+[`plutip-server`](../plutip-server) allows only once active cluster at a time, but nothing stops you from setting up multiple CTL environments and multiple `plutip-server`s by running tests in separate fibers and thus using multiple Plutip clusters simultaneously.
```

Comment:
This is not true. It uses a set of pre-defined ports, so parallel clusters will fail to start unless the ports are explicitly made different.

Reply:
Internally Plutip uses `cardano-wallet` function `randomUnusedTCPPorts` to get a list of ports later assigned to nodes: https://github.com/input-output-hk/cardano-wallet/blob/af82118b5cd5addc60c68dc4fdaf59cb1d228be7/lib/wallet/src/Cardano/Wallet/Network/Ports.hs#L119

So you can run parallel clusters, but nodes in different clusters can be assigned the same port due to race condition, and then there'll be another race condition where one of the nodes will fail to start.
I'll rephrase and mention this in the doc.

Reply:
Ah, I see, yes this is a problem then, will mention this

Reply:
Fixed